lib/class-wp-theme-json-resolver.php

```diff
@@ -71,6 +71,85 @@ private static function get_from_file( $file_path ) {
+	 * containing the a translatable path from theme.json and an array
+	 * of properties that are translatable.
+	 *
+	 * @param array $file_structure_partial A part of a theme.json i18n tree.
+	 * @param array $current_path           An array with a path on the theme.json i18n tree.
+	 *
+	 * @return array An array of arrays each one containing a translatable path and an array of properties that are translatable.
+	 */
+	private static function theme_json_i18_file_structure_to_paths( $file_structure_partial, $current_path = array() ) {
+		$result = array();
+		foreach ( $file_structure_partial as $property => $partial_child ) {
+			if ( is_numeric( $property ) ) {
+				return array(
+					array(
+						'path'              => $current_path,
+						'translatable_keys' => $file_structure_partial,
+					),
+				);
+			}
+			$result = array_merge(
+				$result,
+				self::theme_json_i18_file_structure_to_paths( $partial_child, array_merge( $current_path, array( $property ) ) )
+			);
+		}
+		return $result;
+	}
+
+	/**
+	 * Returns a data structure used in theme.json translation.
+	 *
+	 * @return array An array of theme.json paths that are translatable and the keys that are translatable
+	 */
+	private static function get_theme_json_i18n() {
+		static $theme_json_i18n = null;
+		if ( null === $theme_json_i18n ) {
+			$file_structure  = self::get_from_file( __DIR__ . '/experimental-i18n-theme.json' );
+			$theme_json_i18n = self::theme_json_i18_file_structure_to_paths( $file_structure );
+
+		}
+		return $theme_json_i18n;
+	}
+
+	/**
+	 * Translates a theme.json structure.
+	 *
+	 * @param array  $theme_json_structure A theme.json structure that is going to be translatable.
+	 * @param string $domain               Optional. Text domain. Unique identifier for retrieving translated strings.
+	 *                                     Default 'default'.
+	 */
+	private static function apply_theme_json_translations( &$theme_json_structure, $domain = 'default' ) {
+		$theme_json_i18n = self::get_theme_json_i18n();
+		foreach ( $theme_json_structure as &$context_value ) {
+			if ( empty( $context_value ) ) {
+				continue;
+			}
+			foreach ( $theme_json_i18n as $theme_json_i18n_value ) {
+				$path               = $theme_json_i18n_value['path'];
+				$translatable_keys  = $theme_json_i18n_value['translatable_keys'];
+				$array_to_translate = gutenberg_experimental_get( $context_value, $path, null );
+				if ( null === $array_to_translate ) {
+					continue;
+				}
+				foreach ( $array_to_translate as &$item_to_translate ) {
+					foreach ( $translatable_keys as $translatable_key ) {
+						if ( empty( $item_to_translate[ $translatable_key ] ) ) {
+							continue;
+						}
+						// phpcs:ignore WordPress.WP.I18n.LowLevelTranslationFunction,WordPress.WP.I18n.NonSingularStringLiteralText,WordPress.WP.I18n.NonSingularStringLiteralDomain
+						$item_to_translate[ $translatable_key ] = translate( $item_to_translate[ $translatable_key ], $domain );
```

Comment:
Why do we use `translate` instead of `__`, etc?

The [docs](https://developer.wordpress.org/reference/functions/translate/) say: "don't use translate directly". I haven't seen `translate` used anywhere else in the plugin, we seem to prefer `__` instead.

Reply:
:thinking: I've taken a look at the [code](https://core.trac.wordpress.org/browser/tags/5.6/src/wp-includes/l10n.php#L288) for `__()`:

```php
function __( $text, $domain = 'default' ) {
	return translate( $text, $domain );
}
```

so, essentially, `__` is a direct call to `translate`. I'm not sure why `__` is preferred rather than `translate`, but looking at the code, both have the same behavior. Hence, I don't understand why we can't use `__` here (and so follow the recommendation). Not a big deal, just want to understand this better.

Reply:
From the code point of view, both functions do exactly the same thing.

We have a parser that parses the PHP code and generates the translation file. To extract the strings that are translatable that parser extracts things inside function __(' ... ' ), and we don't want the extraction to happen here. That's the reason why I'm not using `__`.

Reply:
Ah, I think I got it now. The idea is that there's a tool that parses the WordPress PHP code to extract the strings to be translated. That tool uses the `_` function as a _marker_: whenever it encounters it, it registers its strings as something to translate. However, `translate` is not used by that tool as a _marker_. If that's correct, TIL something new! Thanks for keeping up with me.

Reply:
Hi @nosolosw, exactly your comment summarises everything correctly :)